File: ios/HealthTracker/Services/MealAnalysisService.swift
import Foundation
import UIKit
import CoreData

// MARK: - Models

struct MealAnalysis: Codable {
    let items: [DetectedFood]
    let totalCalories: Int
    let confidence: Double
    let imageId: String?
}

struct DetectedFood: Codable, Identifiable {
    let id = UUID()
    var name: String
    var quantity: String
    var calories: Double
    var protein: Double
    var carbs: Double
    var fat: Double
    let fiber: Double?
    let confidence: Double
    let boundingBox: CGRect?

    enum CodingKeys: String, CodingKey {
        case name, quantity, calories, protein, carbs, fat, fiber, confidence, boundingBox
    }
}

// MARK: - API Service

class MealAnalysisService: ObservableObject {
    static let shared = MealAnalysisService()

    @Published var isAnalyzing = false
    @Published var lastAnalysis: MealAnalysis?

    // Meal-scan goes through the Vercel proxy at mochasmindlab.com/api/v1/meal-scan.
    // The Anthropic key lives in the proxy's env vars, not in the app binary.
    // Apps authenticate to the proxy with APP_SHARED_SECRET + a per-install UUID.
    private let proxyEndpoint = SecretsManager.mealScanEndpoint
    private let appSharedSecret = SecretsManager.appSharedSecret

    private init() {}

    // MARK: - Main Analysis Function

    func analyzeMealPhoto(_ image: UIImage, completion: @escaping (Result<MealAnalysis, Error>) -> Void) {
        isAnalyzing = true

        // Resize image to optimize API calls
        let resizedImage = resizeImage(image, maxDimension: 1024)

        guard let imageData = resizedImage.jpegData(compressionQuality: 0.8) else {
            completion(.failure(AnalysisError.imageProcessingFailed))
            isAnalyzing = false
            return
        }

        let base64String = imageData.base64EncodedString()

        // Call Claude Vision API
        analyzeWithClaude(base64Image: base64String, completion: completion)
    }

    // MARK: - Claude Vision API

    private func analyzeWithClaude(base64Image: String, completion: @escaping (Result<MealAnalysis, Error>) -> Void) {
        // Proxy expects a minimal body — image only. Prompt + model selection
        // are owned by the proxy so they can be tuned server-side without app
        // updates. See Web-Projects/mochamindlabs-website/api/v1/meal-scan.js.
        let requestBody: [String: Any] = [
            "image": base64Image
        ]

        guard let url = URL(string: proxyEndpoint) else {
            completion(.failure(AnalysisError.networkError))
            return
        }
        guard !appSharedSecret.isEmpty else {
            completion(.failure(AnalysisError.apiKeyMissing))
            isAnalyzing = false
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue(appSharedSecret, forHTTPHeaderField: "X-App-Secret")
        request.setValue(SecretsManager.installId, forHTTPHeaderField: "X-Install-Id")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: requestBody)

        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            DispatchQueue.main.async {
                self?.isAnalyzing = false

                if let error = error {
                    completion(.failure(error))
                    return
                }

                guard let data = data else {
                    completion(.failure(AnalysisError.noData))
                    return
                }

                // Map proxy HTTP status to friendly errors before attempting to
                // parse — a 401/429 body won't look like an Anthropic response.
                if let http = response as? HTTPURLResponse {
                    switch http.statusCode {
                    case 200...299:
                        break
                    case 401:
                        completion(.failure(AnalysisError.apiKeyMissing))
                        return
                    case 429:
                        completion(.failure(AnalysisError.quotaExceeded))
                        return
                    default:
                        completion(.failure(AnalysisError.invalidResponse))
                        return
                    }
                }

                do {
                    let result = try self?.parseClaudeResponse(data)
                    if let analysis = result {
                        self?.lastAnalysis = analysis
                        completion(.success(analysis))
                    }
                } catch {
                    print("Parse error: \(error)")
                    completion(.failure(error))
                }
            }
        }.resume()
    }

    // MARK: - Response Parsing

    private func parseClaudeResponse(_ data: Data) throws -> MealAnalysis {
        // Debug: print raw response
        if let rawString = String(data: data, encoding: .utf8) {
            print("🔍 Claude raw response:\n\(rawString)")
        }

        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]

        // Check for API error
        if let errorInfo = json?["error"] as? [String: Any],
           let errorMessage = errorInfo["message"] as? String {
            print("❌ Claude API Error: \(errorMessage)")
            throw AnalysisError.invalidResponse
        }

        guard let content = json?["content"] as? [[String: Any]],
              let firstContent = content.first,
              let text = firstContent["text"] as? String else {
            print("❌ Could not extract text from Claude response. Keys: \(json?.keys.joined(separator: ", ") ?? "none")")
            throw AnalysisError.invalidResponse
        }

        // Extract JSON from the response (Claude might include extra text)
        print("📝 Claude text response:\n\(text)")
        let jsonString = extractJSON(from: text)
        print("📦 Extracted JSON:\n\(jsonString)")

        guard let jsonData = jsonString.data(using: .utf8) else {
            throw AnalysisError.invalidResponse
        }

        do {
            let analysis = try JSONDecoder().decode(MealAnalysis.self, from: jsonData)
            return analysis
        } catch {
            print("❌ Decode error: \(error)")
            throw error
        }
    }

    // Helper to extract JSON from Claude's response
    private func extractJSON(from text: String) -> String {
        // Try to find JSON object in the response
        if let startIndex = text.firstIndex(of: "{"),
           let endIndex = text.lastIndex(of: "}") {
            return String(text[startIndex...endIndex])
        }
        return text
    }

    // MARK: - Helper Functions

    private func resizeImage(_ image: UIImage, maxDimension: CGFloat) -> UIImage {
        let size = image.size
        let aspectRatio = size.width / size.height

        var newSize: CGSize
        if size.width > size.height {
            newSize = CGSize(width: maxDimension, height: maxDimension / aspectRatio)
        } else {
            newSize = CGSize(width: maxDimension * aspectRatio, height: maxDimension)
        }

        UIGraphicsBeginImageContextWithOptions(newSize, false, 1.0)
        image.draw(in: CGRect(origin: .zero, size: newSize))
        let resizedImage = UIGraphicsGetImageFromCurrentImageContext()!
        UIGraphicsEndImageContext()

        return resizedImage
    }

    // MARK: - Save to Core Data

    func saveAnalysisToFoodEntry(analysis: MealAnalysis, image: UIImage?, mealType: MealType, context: NSManagedObjectContext) throws {
        try saveAnalysisToFoodEntry(items: analysis.items, mealType: mealType, context: context)
    }

    func saveAnalysisToFoodEntry(items: [DetectedFood], mealType: MealType, context: NSManagedObjectContext) throws {
        for item in items {
            let entry = FoodEntry(context: context)
            entry.id = UUID()
            entry.name = item.name
            entry.calories = item.calories
            entry.protein = item.protein
            entry.carbs = item.carbs
            entry.fat = item.fat
            entry.fiber = item.fiber ?? 0
            entry.timestamp = Date()
            entry.date = Date()
            entry.servingSize = item.quantity
            entry.servingUnit = "serving"
            entry.mealType = mealType.rawValue
        }

        try context.save()
        UnifiedDataManager.shared.refreshAllData()
        ReviewRequestManager.shared.recordMealScanned()
    }
}

// MARK: - Error Types

enum AnalysisError: LocalizedError {
    case imageProcessingFailed
    case noData
    case invalidResponse
    case apiKeyMissing
    case quotaExceeded
    case networkError

    var errorDescription: String? {
        switch self {
        case .imageProcessingFailed:
            return "Failed to process image"
        case .noData:
            return "No data received from server"
        case .invalidResponse:
            return "Invalid response format"
        case .apiKeyMissing:
            return "API key not configured"
        case .quotaExceeded:
            return "API quota exceeded"
        case .networkError:
            return "Network connection error"
        }
    }
}

File: android/app/src/main/java/com/mochasmindlab/mlhealth/services/MealAnalysisService.kt
package com.mochasmindlab.mlhealth.services

import android.graphics.Bitmap
import android.util.Base64
import com.mochasmindlab.mlhealth.data.database.FoodDao
import com.mochasmindlab.mlhealth.data.entities.FoodEntry
import com.mochasmindlab.mlhealth.data.models.MealAnalysis
import com.mochasmindlab.mlhealth.utils.PreferencesManager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.ByteArrayOutputStream
import java.util.*
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

sealed class AnalysisError(message: String) : Exception(message) {
    object ImageProcessingFailed : AnalysisError("Failed to process image")
    object NoData : AnalysisError("No data received from server")
    object InvalidResponse : AnalysisError("Invalid response format")
    object ApiKeyMissing : AnalysisError("App is misconfigured. Please update.")
    object NetworkError : AnalysisError("Network connection error")
    object RateLimited : AnalysisError("Too many scans this hour. Try again later.")
}

@Singleton
class MealAnalysisService @Inject constructor(
    private val preferencesManager: PreferencesManager
) {

    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .build()

    suspend fun analyzeMealPhoto(bitmap: Bitmap): Result<MealAnalysis> =
        withContext(Dispatchers.IO) {
            // Auth: shared secret identifies us as the genuine app to the proxy.
            // Without it the proxy returns 401. (We don't carry an Anthropic key.)
            val sharedSecret = SecretsManager.appSharedSecret
                ?: return@withContext Result.failure(AnalysisError.ApiKeyMissing)

            val installId = preferencesManager.getOrCreateInstallId()

            val base64Image = try {
                encodeImage(bitmap)
            } catch (e: Exception) {
                return@withContext Result.failure(AnalysisError.ImageProcessingFailed)
            }

            try {
                val rawJson = callMealScanProxy(base64Image, sharedSecret, installId)
                val analysis = parseResponse(rawJson)
                Result.success(analysis)
            } catch (e: AnalysisError) {
                Result.failure(e)
            } catch (e: Exception) {
                Result.failure(AnalysisError.NetworkError)
            }
        }

    suspend fun saveAnalysisToDiary(
        analysis: MealAnalysis,
        mealType: String,
        foodDao: FoodDao
    ): Int = withContext(Dispatchers.IO) {
        val today = Date()
        var saved = 0
        for (item in analysis.items) {
            val entry = FoodEntry(
                id = UUID.randomUUID(),
                name = item.name,
                date = today,
                timestamp = today,
                mealType = mealType.lowercase(),
                servingSize = item.quantity,
                servingUnit = "serving",
                servingCount = 1.0,
                calories = item.calories,
                protein = item.protein,
                carbs = item.carbs,
                fat = item.fat,
                fiber = item.fiber
            )
            foodDao.insert(entry)
            saved++
        }
        saved
    }

    private fun encodeImage(bitmap: Bitmap): String {
        val scaled = resizeBitmap(bitmap, 1024)
        val stream = ByteArrayOutputStream()
        scaled.compress(Bitmap.CompressFormat.JPEG, 80, stream)
        val bytes = stream.toByteArray()
        if (bytes.isEmpty()) throw AnalysisError.ImageProcessingFailed
        return Base64.encodeToString(bytes, Base64.NO_WRAP)
    }

    private fun resizeBitmap(src: Bitmap, maxDimension: Int): Bitmap {
        val w = src.width
        val h = src.height
        if (w <= maxDimension && h <= maxDimension) return src
        val scale = maxDimension.toFloat() / maxOf(w, h)
        return Bitmap.createScaledBitmap(src, (w * scale).toInt(), (h * scale).toInt(), true)
    }

    private fun callMealScanProxy(
        base64Image: String,
        sharedSecret: String,
        installId: String
    ): String {
        // Body matches the proxy contract — see Web-Projects/mochamindlabs-website/api/v1/meal-scan.js.
        // The proxy builds the actual Anthropic request server-side so we never
        // ship the Anthropic API key. Prompt + model selection are server-owned too.
        val requestBody = JsonObject(mapOf("image" to JsonPrimitive(base64Image))).toString()
        val request = Request.Builder()
            .url(SecretsManager.mealScanEndpoint)
            .addHeader("X-App-Secret", sharedSecret)
            .addHeader("X-Install-Id", installId)
            // Platform header — proxy uses this to pick the right Anthropic key
            // (when ANTHROPIC_API_KEY_ANDROID is set) and for per-platform logs.
            .addHeader("X-Platform", "android")
            .post(requestBody.toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()
        val body = response.body?.string() ?: throw AnalysisError.NoData
        when (response.code) {
            in 200..299 -> return body
            401 -> throw AnalysisError.ApiKeyMissing
            429 -> throw AnalysisError.RateLimited
            else -> throw AnalysisError.InvalidResponse
        }
    }

    private fun parseResponse(rawJson: String): MealAnalysis {
        val root = try {
            json.parseToJsonElement(rawJson) as? JsonObject
                ?: throw AnalysisError.InvalidResponse
        } catch (e: AnalysisError) {
            throw e
        } catch (e: Exception) {
            throw AnalysisError.InvalidResponse
        }

        if (root.containsKey("error")) throw AnalysisError.InvalidResponse

        val text = root["content"]
            ?.jsonArray
            ?.firstOrNull()
            ?.jsonObject
            ?.get("text")
            ?.jsonPrimitive
            ?.content
            ?: throw AnalysisError.InvalidResponse

        val extracted = extractJson(text)
        return try {
            json.decodeFromString(MealAnalysis.serializer(), extracted)
        } catch (e: Exception) {
            throw AnalysisError.InvalidResponse
        }
    }

    private fun extractJson(text: String): String {
        val start = text.indexOf('{')
        val end = text.lastIndexOf('}')
        if (start == -1 || end == -1 || end < start) return text
        return text.substring(start, end + 1)
    }
}
